ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1627 @@
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAM_DATA_UNI,
+                                                   ossl_quic_rxfc_get_cwm(&ch->stream0->rxfc)))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAMS_BIDI,
+                                                   0))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAMS_UNI,
+                                                   0))
+        goto err;
+
+    if (!WPACKET_get_total_written(&wpkt, &buf_len))
+        goto err;
+
+    ch->local_transport_params = (unsigned char *)buf_mem->data;
+    buf_mem->data = NULL;
+
+    if (!WPACKET_finish(&wpkt))
+        goto err;
+
+    wpkt_valid = 0;
+
+    if (!ossl_quic_dhs_set_transport_params(ch->dhs, ch->local_transport_params,
+                                            buf_len))
+        goto err;
+
+    ok = 1;
+err:
+    if (wpkt_valid)
+        WPACKET_cleanup(&wpkt);
+    BUF_MEM_free(buf_mem);
+    return ok;
+}
+
+/*
+ * QUIC Channel: Ticker-Mutator
+ * ============================
+ */
+
+/*
+ * The central ticker function called by the reactor. This does everything, or
+ * at least everything network I/O related. Best effort - not allowed to fail
+ * "loudly".
+ */
+static void ch_tick(QUIC_TICK_RESULT *res, void *arg)
+{
+    OSSL_TIME now, deadline;
+    QUIC_CHANNEL *ch = arg;
+
+    /*
+     * When we tick the QUIC connection, we do everything we need to do
+     * periodically. In order, we:
+     *
+     *   - handle any incoming data from the network;
+     *   - handle any timer events which are due to fire (ACKM, etc.)
+     *   - write any data to the network due to be sent, to the extent
+     *     possible;
+     *   - determine the time at which we should next be ticked.
+     */
+
+    /* If we are in the TERMINATED state, there is nothing to do. */
+    if (ossl_quic_channel_is_terminated(ch)) {
+        res->want_net_read  = 0;
+        res->want_net_write = 0;
+        res->tick_deadline  = ossl_time_infinite();
+        return;
+    }
+
+    /*
+     * If we are in the TERMINATING state, check if the terminating timer has
+     * expired.
+     */
+    if (ossl_quic_channel_is_terminating(ch)) {
+        now = ossl_time_now();
+
+        if (ossl_time_compare(now, ch->terminate_deadline) >= 0) {
+            ch_on_terminating_timeout(ch);
+            res->want_net_read  = 0;
+            res->want_net_write = 0;
+            res->tick_deadline  = ossl_time_infinite();
+            return; /* abort normal processing, nothing to do */
+        }
+    }
+
+    /* Handle any incoming data from the network. */
+    ch_rx(ch);
+
+    /*
+     * Allow the handshake layer to check for any new incoming data and generate
+     * new outgoing data.
+     */
+    ossl_quic_dhs_tick(ch->dhs);
```

Comment:
If we update the rx secrets here then packets that were previously deferred are now available. However because we don't call ch_rx again we don't spot this. We then end up finishing the ch_tick with work we can still do but instead we go and sit polling the fds waiting for more packets.

Reply:
Good catch!